nexus/saga-recovery/src/recovery.rs

```diff
@@ -0,0 +1,705 @@
+        mut running_sagas_found: BTreeMap<SagaId, nexus_db_model::Saga>,
+    ) -> Plan {
+        let mut builder = PlanBuilder::new(log);
+
+        // First of all, remove finished sagas from our "ignore" set.
+        //
+        // `previously_maybe_done` was computed the last time we ran and
+        // contains sagas that either just started or already finished.  We
+        // couldn't tell.  All we knew is that they were running in-memory but
+        // were not included in our database query for in-progress sagas.  At
+        // this point, though, we've done a second database query for
+        // in-progress sagas.  Any items that aren't in that list either cannot
+        // still be running, so we can safely remove them from our ignore set.
+        let previously_maybe_done = &rest_state.remove_next;
+        for saga_id in previously_maybe_done {
+            if !running_sagas_found.contains_key(saga_id) {
+                builder.saga_infer_done(*saga_id);
+            }
+        }
+
+        // Figure out which of the candidate sagas can clearly be skipped.
+        // Correctness here requires that the caller has already updated the set
+        // of sagas that we're ignoring to include any that may have been
+        // created up to the beginning of the database query.  (They do that by
+        // doing the database query first and then updating this set.)  Since we
+        // now have the list of sagas that were not-finished in the database, we
+        // can compare these two sets.
+        let sagas_started = &rest_state.sagas_started;
+        for running_saga_id in sagas_started.keys() {
+            match running_sagas_found.remove(running_saga_id) {
+                None => {
+                    // If this saga is in `previously_maybe_done`, then we
+                    // processed it above already.  We know it's done.
+                    //
+                    // Otherwise, the saga is in the ignore set, but not the
+                    // database list of running sagas.  It's possible that the
+                    // saga has simply finished.  And if the saga is definitely
+                    // not running any more, then we can remove it from the
+                    // ignore set.  This is important to keep that set from
+                    // growing without bound.
+                    //
+                    // But it's also possible that the saga started immediately
+                    // after the database query's snapshot, in which case we
+                    // don't really know if it's still running.
+                    //
+                    // The way to resolve this is to do another database query
+                    // for unfinished sagas.  If it's not in that list, the saga
+                    // must have finished.  Rather than do that now, we'll just
+                    // keep track of this list and take care of it on the next
+                    // pass.
```

Comment:
I think this is a really nice solution for how to resolve the potential race. Makes sense to me!